modules/apps/headless/headless-builder/headless-builder-test/src/testIntegration/java/com/liferay/headless/builder/publisher/test/APIApplicationPublisherTest.java

```diff
@@ -0,0 +1,311 @@
+import com.liferay.batch.engine.unit.BatchEngineUnitProcessor;
+import com.liferay.batch.engine.unit.BatchEngineUnitReader;
+import com.liferay.headless.builder.application.APIApplication;
+import com.liferay.headless.builder.application.provider.APIApplicationProvider;
+import com.liferay.headless.builder.application.provider.test.APIApplicationProviderTest;
+import com.liferay.headless.builder.application.publisher.APIApplicationPublisher;
+import com.liferay.portal.kernel.json.JSONUtil;
+import com.liferay.portal.kernel.test.rule.AggregateTestRule;
+import com.liferay.portal.kernel.test.util.HTTPTestUtil;
+import com.liferay.portal.kernel.test.util.TestPropsValues;
+import com.liferay.portal.kernel.util.Http;
+import com.liferay.portal.test.rule.Inject;
+import com.liferay.portal.test.rule.LiferayIntegrationTestRule;
+
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import javax.ws.rs.core.Application;
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.ServiceReference;
+import org.osgi.util.tracker.ServiceTracker;
+
+/**
+ * @author Luis Miguel Barcos
+ */
+@RunWith(Arquillian.class)
+public class APIApplicationPublisherTest {
+
+	@ClassRule
+	@Rule
+	public static final AggregateTestRule aggregateTestRule =
+		new LiferayIntegrationTestRule();
+
+	@Before
+	public void setUp() {
+		Bundle testBundle = FrameworkUtil.getBundle(
+			APIApplicationProviderTest.class);
+
+		BundleContext bundleContext = testBundle.getBundleContext();
+
+		for (Bundle bundle : bundleContext.getBundles()) {
+			if (Objects.equals(
+					bundle.getSymbolicName(),
+					"com.liferay.headless.builder.impl")) {
+
+				CompletableFuture<Void> completableFuture =
+					_batchEngineUnitProcessor.processBatchEngineUnits(
+						_batchEngineUnitReader.getBatchEngineUnits(bundle));
+
+				completableFuture.join();
+			}
+		}
+	}
+
+	@Test
+	public void testPublishAPIApplication() throws Exception {
+		Bundle bundle = FrameworkUtil.getBundle(getClass());
+
+		BundleContext bundleContext = bundle.getBundleContext();
+
+		CountDownLatch addedCountLatch = new CountDownLatch(1);
+		CountDownLatch removedCountLatch = new CountDownLatch(1);
+
+		ServiceTracker<?, ?> serviceTracker =
+			new ServiceTracker<Application, Application>(
+				bundleContext, Application.class, null) {
+
+				@Override
+				public Application addingService(
+					ServiceReference<Application> serviceReference) {
+
+					Boolean property = (Boolean)serviceReference.getProperty(
+						"liferay.headless.builder.application");
+
+					if ((property != null) && property) {
```

Comment:
We could use GetterUtil and avoid two conditions:

```
if (GetterUtil.getBoolean(
		serviceReference.getProperty(
			"liferay.headless.builder.application"))) {

	addedCountLatch.countDown();

	return super.addingService(serviceReference);
}
```

Reply:
Done